src/Lean/Server/FileWorker/LspEncoding.lean

```diff
@@ -0,0 +1,239 @@
+
+    @[implementedBy encodeUnsafe]
+    private constant encode (r : WithRpcRef $typeId:ident) : RequestM Lsp.RpcRef
+
+    private unsafe def decodeUnsafe (r : Lsp.RpcRef) : RequestM (WithRpcRef $typeId:ident) :=
+      WithRpcRef.decodeUnsafeAs $typeId:ident $(quote typeNm) r
+
+    @[implementedBy decodeUnsafe]
+    private constant decode (r : Lsp.RpcRef) : RequestM (WithRpcRef $typeId:ident)
+
+    instance : LspEncoding (WithRpcRef $typeId:ident) Lsp.RpcRef where
+      lspEncode a := encode a
+      lspDecode a := decode a
+    end $typeId:ident
+  )
+  elabCommand cmds
+  return true
+
+-- TODO remove
+elab "mkWithRefInstance" typeId:ident : command => do let _ ← deriveWithRefInstance typeId.getId
+
+/-- Creates an `LspPacket` structure for `α` with all fields of `α` replaced by their `LspEncoding`
+and uses that to instantiate `LspEncoding α LspPacket`. -/
+private def deriveInstance (typeName : Name) : CommandElabM Bool := do
+  let env ← getEnv
+  if !(← isStructure env typeName) then
+    throwError "only structures are supported"
+  let indVal ← getConstInfoInduct typeName
+  if indVal.all.length ≠ 1 then
+    throwError "mutual induction is unsupported"
+
+  let ctorVal ← getConstInfoCtor indVal.ctors.head!
+  -- TODO(WN): helper to get flattened fields *and* their types
+  let fields := getStructureFields env typeName
+  let numParams := indVal.numParams
+  let cmds : Syntax ← liftTermElabM none do
+    -- Return the `β`, if any, in `LspEncoding tp β`.
+    let hasLspEncoding? (tp : Expr) : TermElabM (Option Expr) := do
+      let clTp ←
+        try
+          let β ← mkFreshExprMVar (mkSort levelOne)
+          mkAppM ``LspEncoding #[tp, β]
+        catch ex =>
+          throwError "failed to construct 'LspEncoding {tp} ?_':\n{ex.toMessageData}"
+      match (← trySynthInstance clTp) with
+      | LOption.some e =>
+        let eT ← inferType e
+        if !eT.isAppOfArity ``LspEncoding 2 then
+          throwError "expected 'LspEncoding {tp} ?_', got '{eT}'"
+        some eT.appArg!
+      | _ => none
+
+    forallTelescopeReducing ctorVal.type fun xs _ => do
+      if xs.size != numParams + fields.size then
+        throwError "unexpected number of fields in structure"
+      let mut fieldEncTs := #[]
+      for i in [:fields.size] do
+        let field := fields[i]
+        let x := xs[numParams + i]
+        let fieldT ← inferType x
+        let some fieldEncT ← hasLspEncoding? fieldT
+          | throwError "cannot synthesize 'LspEncoding {fieldT} ?_'"
+        -- TODO(WN): delab instead of using const name
+        let some fieldEncTName ← fieldEncT.constName?
+          | throwError "type '{fieldEncT}' is not constant"
+        fieldEncTs := fieldEncTs.push fieldEncTName
+
+      let typeId := mkIdent typeName
+
+      let fieldIds := fields.map mkIdent
+      let fieldEncTIds := fieldEncTs.map mkIdent
+
+      -- HACK(WN): the syntax this produces looks royally messed up but elaborates fine
```

Comment:
The `trace.Elab.command` output is wonky, but it may actually just be broken formatting of hygienic idents:
```lean
[Elab.command] instance  : LspEncoding✝ (WithRpcRef✝ FooRef) Lsp.RpcRef✝where lspEncode✝ a✝ := encode✝ a✝lspDecode✝
  a✝ := decode✝ a✝
```
It's not high prio. I'll remove the comment.